File: python/src/adapters/registration/kiro/sso_device_auth.py
"""SSO Device Auth — exchange sso_token for refreshToken + accessToken.

Faithfully ported from Kiro-auto-register/src/main/index.ts ssoDeviceAuth.
7-step flow: register client → device auth → whoAmI → session → accept → approve → poll token.
"""

from __future__ import annotations

import logging
import time

import httpx

from .constants import OIDC_BASE, SCOPES, START_URL

log = logging.getLogger(__name__)


def sso_device_auth(bearer_token: str, region: str = "us-east-1") -> dict:
    """Execute the full SSO device authorization flow.

    Args:
        bearer_token: The x-amz-sso_authn cookie value.
        region: AWS region for OIDC endpoint.

    Returns:
        Dict with keys: success, accessToken, refreshToken, clientId, clientSecret, region, expiresIn, error.
    """
    oidc_base = f"https://oidc.{region}.amazonaws.com"
    portal_base = "https://portal.sso.us-east-1.amazonaws.com"

    with httpx.Client(timeout=30) as client:
        # Step 1: Register OIDC client
        log.info("[SSO] Step 1: Registering OIDC client")
        resp = client.post(f"{oidc_base}/client/register", json={
            "clientName": "Kiro Account Manager",
            "clientType": "public",
            "scopes": SCOPES,
            "grantTypes": ["urn:ietf:params:oauth:grant-type:device_code", "refresh_token"],
            "issuerUrl": START_URL,
        })
        if resp.status_code != 200:
            return _fail(f"Register client failed: {resp.status_code} {resp.text[:200]}")
        reg = resp.json()
        client_id = reg["clientId"]
        client_secret = reg["clientSecret"]

        # Step 2: Device authorization
        log.info("[SSO] Step 2: Starting device authorization")
        resp = client.post(f"{oidc_base}/device_authorization", json={
            "clientId": client_id,
            "clientSecret": client_secret,
            "startUrl": START_URL,
        })
        if resp.status_code != 200:
            return _fail(f"Device auth failed: {resp.status_code} {resp.text[:200]}")
        dev = resp.json()
        device_code = dev["deviceCode"]
        user_code = dev["userCode"]
        interval = dev.get("interval", 1)

        # Step 3: Verify bearer token
        log.info("[SSO] Step 3: Verifying bearer token")
        resp = client.get(f"{portal_base}/token/whoAmI", headers={
            "Authorization": f"Bearer {bearer_token}",
            "Accept": "application/json",
        })
        if resp.status_code != 200:
            return _fail(f"whoAmI failed: {resp.status_code} {resp.text[:200]}")

        # Step 4: Get device session token
        log.info("[SSO] Step 4: Getting device session token")
        resp = client.post(f"{portal_base}/session/device", headers={
            "Authorization": f"Bearer {bearer_token}",
            "Content-Type": "application/json",
        }, json={})
        if resp.status_code != 200:
            return _fail(f"Device session failed: {resp.status_code} {resp.text[:200]}")
        device_session_token = resp.json()["token"]

        # Step 5: Accept user code
        log.info("[SSO] Step 5: Accepting user code %s", user_code)
        resp = client.post(f"{oidc_base}/device_authorization/accept_user_code", headers={
            "Content-Type": "application/json",
            "Referer": "https://view.awsapps.com/",
        }, json={
            "userCode": user_code,
            "userSessionId": device_session_token,
        })
        if resp.status_code != 200:
            return _fail(f"Accept user code failed: {resp.status_code} {resp.text[:200]}")
        accept_data = resp.json()
        device_context = accept_data.get("deviceContext")

        # Step 6: Approve authorization
        if device_context and device_context.get("deviceContextId"):
            log.info("[SSO] Step 6: Approving authorization")
            resp = client.post(f"{oidc_base}/device_authorization/associate_token", headers={
                "Content-Type": "application/json",
                "Referer": "https://view.awsapps.com/",
            }, json={
                "deviceContext": {
                    "deviceContextId": device_context["deviceContextId"],
                    "clientId": device_context.get("clientId", client_id),
                    "clientType": device_context.get("clientType", "public"),
                },
                "userSessionId": device_session_token,
            })
            if resp.status_code != 200:
                return _fail(f"Approve failed: {resp.status_code} {resp.text[:200]}")

        # Step 7: Poll for token
        log.info("[SSO] Step 7: Polling for token")
        deadline = time.time() + 120

        while time.time() < deadline:
            time.sleep(interval)
            resp = client.post(f"{oidc_base}/token", json={
                "clientId": client_id,
                "clientSecret": client_secret,
                "grantType": "urn:ietf:params:oauth:grant-type:device_code",
                "deviceCode": device_code,
            })

            if resp.status_code == 200:
                token_data = resp.json()
                log.info("[SSO] Token obtained successfully")
                return {
                    "success": True,
                    "accessToken": token_data["accessToken"],
                    "refreshToken": token_data["refreshToken"],
                    "clientId": client_id,
                    "clientSecret": client_secret,
                    "region": region,
                    "expiresIn": token_data.get("expiresIn"),
                }

            if resp.status_code == 400:
                err = resp.json()
                if err.get("error") == "authorization_pending":
                    continue
                if err.get("error") == "slow_down":
                    interval += 5
                    continue
                return _fail(f"Token poll error: {err.get('error')}")

        return _fail("Authorization timeout (120s)")


def _fail(msg: str) -> dict:
    log.error("[SSO] %s", msg)
    return {"success": False, "error": msg}

File: worker-dashboard/src/services/sso-device-auth.ts
/**
 * SSO Device Auth — exchange sso_token for refresh_token + access_token.
 */

import { log } from "../lib/log.js";

interface DeviceAuthResult {
  success: boolean;
  accessToken?: string;
  refreshToken?: string;
  clientId?: string;
  clientSecret?: string;
  region?: string;
  expiresIn?: number;
  error?: string;
}

const SCOPES = [
  "codewhisperer:analysis", "codewhisperer:completions",
  "codewhisperer:conversations", "codewhisperer:taskassist", "codewhisperer:transformations",
];

export async function ssoDeviceAuth(bearerToken: string, region = "us-east-1"): Promise<DeviceAuthResult> {
  const oidcBase = `https://oidc.${region}.amazonaws.com`;
  const portalBase = "https://portal.sso.us-east-1.amazonaws.com";
  const startUrl = "https://view.awsapps.com/start";

  try {
    const regResp = await fetch(`${oidcBase}/client/register`, {
      method: "POST", headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ clientName: "Kiro Account Manager", clientType: "public", scopes: SCOPES, grantTypes: ["urn:ietf:params:oauth:grant-type:device_code", "refresh_token"], issuerUrl: startUrl }),
    });
    if (!regResp.ok) return fail(`Register client: HTTP ${regResp.status}`);
    const reg = await regResp.json() as { clientId: string; clientSecret: string };

    const devResp = await fetch(`${oidcBase}/device_authorization`, {
      method: "POST", headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ clientId: reg.clientId, clientSecret: reg.clientSecret, startUrl }),
    });
    if (!devResp.ok) return fail(`Device auth: HTTP ${devResp.status}`);
    const dev = await devResp.json() as { userCode: string; deviceCode: string; interval?: number };
    let interval = dev.interval || 1;

    const whoResp = await fetch(`${portalBase}/token/whoAmI`, { headers: { Authorization: `Bearer ${bearerToken}`, Accept: "application/json" } });
    if (!whoResp.ok) return fail(`whoAmI: HTTP ${whoResp.status}`);

    const sessResp = await fetch(`${portalBase}/session/device`, { method: "POST", headers: { Authorization: `Bearer ${bearerToken}`, "Content-Type": "application/json" }, body: "{}" });
    if (!sessResp.ok) return fail(`Device session: HTTP ${sessResp.status}`);
    const { token: deviceSessionToken } = await sessResp.json() as { token: string };

    const acceptResp = await fetch(`${oidcBase}/device_authorization/accept_user_code`, {
      method: "POST", headers: { "Content-Type": "application/json", Referer: "https://view.awsapps.com/" },
      body: JSON.stringify({ userCode: dev.userCode, userSessionId: deviceSessionToken }),
    });
    if (!acceptResp.ok) return fail(`Accept user code: HTTP ${acceptResp.status}`);
    const { deviceContext } = await acceptResp.json() as { deviceContext?: { deviceContextId: string; clientId?: string; clientType?: string } };

    if (deviceContext?.deviceContextId) {
      const approveResp = await fetch(`${oidcBase}/device_authorization/associate_token`, {
        method: "POST", headers: { "Content-Type": "application/json", Referer: "https://view.awsapps.com/" },
        body: JSON.stringify({ deviceContext: { deviceContextId: deviceContext.deviceContextId, clientId: deviceContext.clientId || reg.clientId, clientType: deviceContext.clientType || "public" }, userSessionId: deviceSessionToken }),
      });
      if (!approveResp.ok) return fail(`Approve: HTTP ${approveResp.status}`);
    }

    const deadline = Date.now() + 60000;
    while (Date.now() < deadline) {
      await new Promise((r) => setTimeout(r, interval * 1000));
      const tokenResp = await fetch(`${oidcBase}/token`, {
        method: "POST", headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ clientId: reg.clientId, clientSecret: reg.clientSecret, grantType: "urn:ietf:params:oauth:grant-type:device_code", deviceCode: dev.deviceCode }),
      });

      if (tokenResp.ok) {
        const data = await tokenResp.json() as { accessToken: string; refreshToken: string; expiresIn: number };
        log.info("sso_device_auth_ok");
        return { success: true, accessToken: data.accessToken, refreshToken: data.refreshToken, clientId: reg.clientId, clientSecret: reg.clientSecret, region, expiresIn: data.expiresIn };
      }

      if (tokenResp.status === 400) {
        const err = await tokenResp.json() as { error: string };
        if (err.error === "authorization_pending") continue;
        if (err.error === "slow_down") { interval += 5; continue; }
        return fail(`Token poll: ${err.error}`);
      }
    }
    return fail("Authorization timeout (60s)");
  } catch (e) {
    const msg = e instanceof Error ? e.message : "unknown";
    log.error("sso_device_auth_error", { error: msg });
    return fail(msg);
  }
}

function fail(error: string): DeviceAuthResult {
  log.warn("sso_device_auth_failed", { error });
  return { success: false, error };
}
